### PR TITLE
Immediate executor

### PR DESCRIPTION
**Pull request by klemens-morgenstern**

### CODE REVIEW

**Comment by klemens-morgenstern** [comment]
Note: `get_associated_immediate_executor` works with composed ops.

**Comment by klemens-morgenstern** [comment]
Specific tests are still needed. boost.async supports it, so that might be a useful first test.

**Comment by klemens-morgenstern** [comment]
@ashtum Can you review this?

File: include/boost/beast/core/async_base.hpp [hunk 1]
```diff
@@ -309,15 +330,29 @@ class async_base
...
-        If a class derived from @ref async_base is a completion
-        handler, then the associated cancellation_slot of the
-        derived class will be this type.
+    /** Returns the executor associated with this object.
 
-        The default type is a filtering cancellation slot,
-        that only allows terminal cancellation.
+        If a class derived from @ref boost::beast::async_base is a completion
+        handler, then the object returned from this function will be used
+        as the associated executor of the derived class.
```
[the remark this answers is not in the document]
Comment (klemens-morgenstern) [reply include/boost/beast/core/async_base.hpp]: Update the comment

File: include/boost/beast/core/async_base.hpp [hunk 2]
```diff
@@ -309,15 +330,31 @@ class async_base
...
-        that only allows terminal cancellation.
+        If a class derived from @ref boost::beast::async_base is a completion
+        handler, then the object returned from this function will be used
+        as the associated immediate executor of the derived class.
     */
+    immediate_executor_type
+    get_immediate_executor() const noexcept
+    {
+      return net::get_associated_immediate_executor(
+          h_, wg1_.get_executor());
```
Comment (fpelliccioni) [inline include/boost/beast/core/async_base.hpp:345]: NIT: formatting 
```suggestion
          return net::get_associated_immediate_executor(
               h_, wg1_.get_executor());
```

File: include/boost/beast/websocket/impl/read.hpp [hunk 3]
```diff
@@ -291,7 +292,8 @@ class stream<NextLayer, deflateSupported>::read_some_op
                                     __FILE__, __LINE__,
                                     "websocket::async_read_some"));
 
-                                net::post(sp->stream().get_executor(), std::move(*this));
+                                const auto ex = this->get_immediate_executor();
+                    net::dispatch(ex, std::move(*this));
```
Comment (fpelliccioni) [inline include/boost/beast/websocket/impl/read.hpp:296]:
```suggestion
                                const auto ex = this->get_immediate_executor();
                                net::dispatch(ex, std::move(*this));
```

File: include/boost/beast/websocket/impl/read.hpp [hunk 4]
```diff
@@ -335,7 +337,8 @@ class stream<NextLayer, deflateSupported>::read_some_op
                                         __FILE__, __LINE__,
                                         "websocket::async_read_some"));
 
-                                    net::post(sp->stream().get_executor(), std::move(*this));
+                                    const auto ex = this->get_immediate_executor();
+                    net::dispatch(ex, std::move(*this));
```
Comment (fpelliccioni) [inline include/boost/beast/websocket/impl/read.hpp:341]:
```suggestion
                                    const auto ex = this->get_immediate_executor();
                                    net::dispatch(ex, std::move(*this));
```

File: include/boost/beast/websocket/impl/read.hpp [hunk 5]
```diff
@@ -366,7 +369,8 @@ class stream<NextLayer, deflateSupported>::read_some_op
...
```
Comment (fpelliccioni) [inline include/boost/beast/websocket/impl/read.hpp:373]:
```suggestion
                                    const auto ex = this->get_immediate_executor();
                                    net::dispatch(ex, std::move(*this));
```

**Comment by fpelliccioni** [comment]
The changes in the code appear to be consistent and well-structured.

Updates made to the function comments, which generate documentation, are appropriate and coherent.

There don't seem to be significant changes to the API, with the exception of async_base::get_immediate_executor(). Given this minor adjustment, the tests appear to remain effective and relevant.

Codecov has identified several newly added lines of code that aren’t covered by tests. While I’m not entirely certain, these seem to be false positives. It would be beneficial to review and confirm these findings, either by adding the necessary tests or silencing irrelevant warnings. For example:

```
Codecov
/ codecov/patch
include/boost/beast/http/impl/write.hpp#L224-L225
Added lines #L224 - L225 were not covered by tests
```

The PR lacks a description. I recommend adding a brief description to assist both current reviewers and anyone looking at this PR in the future. This description should include the rationale behind the proposed changes (even if it seems obvious), and whether the changes resolve or address any specific issues or problems.

In summary, this PR appears to be mostly well-structured and presented. The most crucial step at this point would be to address the areas identified by Codecov and to enhance the PR description to provide clear context and complete understanding to reviewers and future observers.

Good job 🚀

File: include/boost/beast/core/async_base.hpp [hunk 6]
```diff
@@ -18,10 +18,12 @@
 #include <boost/beast/core/detail/work_guard.hpp>
 #include <boost/asio/associated_allocator.hpp>
 #include <boost/asio/associated_executor.hpp>
+#include <boost/asio/associated_immediate_executor.hpp>
 #include <boost/asio/bind_executor.hpp>
 #include <boost/asio/handler_alloc_hook.hpp>
 #include <boost/asio/handler_continuation_hook.hpp>
 #include <boost/asio/handler_invoke_hook.hpp>
+#include <boost/asio/dispatch.hpp>
 #include <boost/asio/post.hpp>
```
Comment (ashtum) [inline include/boost/beast/core/async_base.hpp:27]: Unused header.
